Add a more flexible GA build to test on Java 8 and 11 inside docker and on mac, lin and win

mathieucarbou opened the pull request:
- to test inside and outside containers
- to test with Java 8 and 11
- to test on lin / win and mac

Made some corrections to allow the build to run on Java 11 (project build was not compatible with Java 11)

**IMPORTANT NOTE:** Github runners do not currently support IPv6, their infrastructure is not routing IPv6. See: https://github.com/actions/virtual-environments/issues/668.

You can see the build output here: https://github.com/mathieucarbou/terracotta-utilities/runs/

![image](https://user-images.githubusercontent.com/61346/154734573-26bf132a-1eac-4552-9ae5-cb5d4862f585.png)

mathieucarbou commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,17 @@
+  push:
+    branches:
+      - master
+      - lsof # just to be able to run all these builds on our own fork
```
Could be removed if we go with GA build

mathieucarbou commented on pom.xml:
```diff
@@ -238,16 +239,16 @@
+<!--            </goals>-->
+<!--          </execution>-->
+<!--        </executions>-->
+<!--      </plugin>-->
```
Toolchain cannot be used easily and anyway obsoleted by the fact that we control what's inside a docker image. So this would need to be moved in a profile that we can disable when running on GA

mathieucarbou commented on pom.xml:
```diff
@@ -366,6 +369,26 @@
+            <artifactId>maven-toolchains-plugin</artifactId>
+            <executions>
+              <execution>
+                <phase>none</phase>
```
option to skip toolchain inside a container build

mathieucarbou commented on port-chooser/pom.xml:
```diff
@@ -39,6 +39,30 @@
+            <id>IPv4</id>
+            <configuration>
+              <systemPropertyVariables>
+                <java.net.preferIPv4Stack>true</java.net.preferIPv4Stack>
```
dual run of the tests with IPv4

mathieucarbou commented on test-tools/src/test/java/org/terracotta/org/junit/rules/TemporaryFolderUsageTest.java:
```diff
@@ -98,6 +99,8 @@ public void newFolderWithGivenFolderThrowsIOExceptionIfFileExists() throws IOExc
 
     @Test
     public void newFolderWithGivenFolderThrowsIOExceptionWhenFolderCannotBeCreated() throws IOException {
+        assumeFalse(Boolean.getBoolean("inside-container"));
```
This use case is not supported inside Docker.
Probably `setReadOnly()` has no effect ?

akomakom commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,69 @@
+          distribution: ${{matrix.distribution}}
+      - name: Test with Maven
+        run: ./mvnw clean verify -V -B -D"maven.artifact.threads=64" -D"org.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn" -D"disable-toolchain" -D"java.version=${{matrix.java}}" --fail-never
+      - name: Upload Unit Test Results
```
Why not publish the test results immediately?

mathieucarbou replied:
https://github.com/marketplace/actions/publish-unit-test-results

mathieucarbou commented:
Updated and rebased the PR, added test results - see screenshot and example: https://github.com/mathieucarbou/terracotta-utilities/runs/5251256039?check_suite_focus=true

mathieucarbou commented:
replaced by #37
